### PR TITLE
Support recipes that and in .yaml as well as .yml

### LINKED ISSUE #11354: YAML recipes are not found unless the extension is `.yml`
## Description
YAML recipes with the extension `.yaml` are not found.

Looks like the offending code is here:

https://github.com/chef/chef/blob/6bf9c2de6c63cfa9b03e61ad5ba02ddb0d8a9b05/lib/chef/cookbook_version.rb#L586

## Chef Version
16.13.16

## Platform Version
ubuntu 20.04

## Replication Case

Create a cookbook with a yaml recipe in the `recipes` directory and try to converge it.

```
my_cookbook
  recipes
    default.yaml
```

## Client Output

```
Compiling Cookbooks...
  [2021-04-12T20:37:40+00:00] DEBUG: Cookbooks to compile: [:my_cookbook]
  [2021-04-12T20:37:40+00:00] TRACE: Loading recipe my_cookbook::default via include_recipe
  [2021-04-12T20:37:40+00:00] TRACE: Filtered backtrace of compile error:

  ================================================================================
  Recipe Compile Error
  ================================================================================

  Chef::Exceptions::RecipeNotFound
  --------------------------------
  could not find recipe default for cookbook my_cookbook

  System Info:
  ------------
  chef_version=16.13.16
  platform=ubuntu
  platform_version=20.04
  ruby=ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-linux]
  program_name=/opt/chef/bin/chef-client
  executable=/opt/chef/bin/chef-client

  [2021-04-12T20:37:40+00:00] TRACE: Re-raising exception: Chef::Exceptions::RecipeNotFound - could not find recipe default for cookbook my_cookbook
  /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/cookbook_version.rb:204:in `load_recipe'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context.rb:393:in `load_recipe'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context/cookbook_compiler.rb:174:in `block in compile_recipes'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context/cookbook_compiler.rb:171:in `each'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context/cookbook_compiler.rb:171:in `compile_recipes'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context/cookbook_compiler.rb:79:in `compile'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/run_context.rb:223:in `load'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/policy_builder/policyfile.rb:196:in `setup_run_context'
   /opt/chef/embedded/lib/ruby/2.7.0/forwardable.rb:235:in `setup_run_context'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/client.rb:496:in `setup_run_context'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/client.rb:282:in `run'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/application.rb:305:in `run_with_graceful_exit_option'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/application.rb:281:in `block in run_chef_client'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/local_mode.rb:42:in `with_server_connectivity'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/application.rb:264:in `run_chef_client'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/application/base.rb:337:in `run_application'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.13.16/lib/chef/application.rb:67:in `run'
   /opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-bin-16.13.16/bin/chef-client:25:in `<top (required)>'
   /opt/chef/bin/chef-client:146:in `load'
   /opt/chef/bin/chef-client:146:in `<main>'
```

### PR DESCRIPTION
**Pull request by marcparadise**
This adds support for recipe files that end in the '.yaml'
extension in addition to existing support for '.yml', because
both are valid and common extensions for yaml files.

Of note is that the message we log when a recipes/default.y[a]ml
and a root_files/recipe.y[a]ml both exist will be slightly off, in that
we don't capture the actual extension of each file. This seems to be
more complexity than it was worth, for a message that is shown at the
beginning of the run and largely ignored.

I also tried to add a bit of localized unit testing for
`recipe_yml_filenames_by_name` because we didn't before.  Mostly focuses
on this use case (yaml, yml both work) but we could probalby improve
that by including tests for a mix of yaml/yml files.

In addition, we will now fail the run when a cookbook contains two same-named 
YAML files with different extensions. THis is because we don't know which one to use - 
either one is a valid YAML file extension. Instead of picking one
arbitrarily we'll fail with an error explaining how to resolve it. 

For example if the cookbook contains both recipes/default.yaml and recipes/default.yml 
it will fail with the following output:
```
Cookbook test1@0.1.0 contains ambiguous files: recipes/default.yaml and recipes/default.yml. Please update the cookbook to remove the incorrect file.
```

Signed-off-by: Marc A. Paradise <marc.paradise@gmail.com>

Fixes #11354

### CODE REVIEW

File: lib/chef/cookbook_version.rb [hunk 1]
```diff
@@ -583,7 +586,7 @@ def filenames_by_name(records)
     end
 
     def yml_filenames_by_name(records)
-      records.select { |record| record[:name] =~ /\.yml$/ }.inject({}) { |memo, record| memo[File.basename(record[:name], ".yml")] = record[:full_path]; memo }
+      records.select { |record| record[:name] =~ /\.(y[a]?ml)$/ }.inject({}) { |memo, record| memo[File.basename(record[:name], File.extname(record[:name]))] = record[:full_path]; memo }
```
Comment (lamont-granquist) [inline lib/chef/cookbook_version.rb:589]: what happens if we have both default.yaml and default.yml?

Comment (marcparadise) [reply lib/chef/cookbook_version.rb:589]: We can either fail the run; or document that one is preferable to the other, and prefer it when both are present. 

I'd rather see us fail the run -- a  de-facto default  like that can easily lead to confusion. 

But I also think we should do the same when default.yml and recipes/recipe.y[a]ml are both present; instead we just log an error and continue.

Comment (marcparadise) [reply lib/chef/cookbook_version.rb]: With the latest updates, we will fail the run with a message in any case we have a same-named file with different yaml/yml extension.

**Comment by marcparadise** [comment]
Note that the lint failure is due to rubocop linting vendored gems.

**Review by johnmccrae — APPROVED** [review]

**Pull request merged**